Fix git_checkout_tree() to do index filemodes correctly on Windows.

jfultz opened the pull request:
git_checkout_tree() has some fallback behaviors for file systems
which don't have full support of filemodes.  Generally works fine,
but if a given file had a change of type from a 0644 to 0755 (i.e.,
you add executable permissions), the fallback behavior incorrectly
triggers when writing hte updated index.

This would cause a git_checkout_tree() command, even with the
GIT_CHECKOUT_FORCE option set, to leave a dirty index on Windows.

Also added checks to an existing test to catch this case.

jfultz commented:
This bug has been thorny in my own libgit2 usage...frequently dirtying the index in a couple of problem repos.  I'd appreciate consideration of porting this to the maintenance release branch for the next v0.22.x release, as well.

carlosmn commented on src/checkout.c:
```diff
@@ -1794,6 +1794,9 @@ static int checkout_create_the_new(
 	size_t i;
+	int caps = git_index_caps(data->index);
+
+	git_index_set_caps(data->index, caps & !GIT_INDEXCAP_NO_FILEMODE);
```
This looks very suspicious. `!GIT_INDEXCAP_NO_FILEMODE` is the logical negation of this flag. As its value is 2, this would make this expression `caps & 0` thus negating all of the caps. You're turning off `NO_FILEMODE` but that's just an accident. What you want to write is likely `caps & (~GIT_INDEXCAP_NO_FILEMODE)` which will pass through any bits set in `caps` except for the `NO_FILEMODE` one.

jfultz replied:
Well, that's just embarrassing.  Fixed.

ethomson commented:
Hey John - sorry about the delay.  I'm not quite back in the office yet and so I'm not in front of my Windows box right now, but in reading this, I'm wondering if the index is actually a little deficient, and perhaps the place to fix this, instead of in checkout.

It looks like `index_merge_mode` tries to sanitize the input mode based on the capabilities.  This seems probably correct when we want to merge the on-disk mode with the mode that exists in the index (which is to say, in `git_index_add_bypath`).  But I think that we *do* want to trust the mode when it is legitimately given to us by the caller (in `git_index_add` and `git_index_add_frombuffer`).

To *not* do so, I think, is another manifestation of this issue, where if you were on a Windows machine and tried to create an index entry with mode `0755` then we would ignore your obviously legitimate request and put it as `0644` in the index.  Disappointing.

I wonder if the right thing to do here isn't to add another flag to `index_insert` that specifies whether the mode should be trusted or not, and only call `index_merge_mode` on untrusted modes.  Then we would not trust the modes from `git_index_add_bypath` but *do* start trusting modes from callers.

Does this seem like the more general fix?

Thanks for all this.  Really appreciate the detailed bug report, test case and a fix!

jfultz commented:
Ed, yeah, I knew you were at Git Merge 2015, and I figured that explained the protracted silence on the pull request.  No worries.

I think this stuff was out of my view because I haven't really used the index API in libgit2 much, yet.  The client I'm developing hasn't needed it until now and (coincidentally) it was the very next thing I was going to be working on.  I understand the basic point you're making, but I want to make sure I really grok the index API before commenting.  So, I'll get back to you in a couple of days...maybe with more code. :)

jfultz commented:
@ethomson I believe this pull request is ready for review again.  After some considerable looking at the code, I agreed with your analysis and implemented it.  I also implemented new tests which show the previous failures in `git_index_add` and `git_index_add_frombuffer` that you point out (and I confirmed that those tests actually did break).

In regards to #2926, this possibly may fix the described behavior.  Unfortunately, I'm unable to reproduce the described behavior because the merge is falling over due to permissions conflicts (i.e., my merge doesn't get far enough to try to change the index because it detects permissions changes in the branch history which it refuses to resolve).  That, um, seems pretty notable to try to resolve, but it's outside the scope of this PR, and it would require significantly more understanding than I presently have of how msysgit (or perhaps git in general...I didn't yet try this elsewhere) deals with permissions issues.

ethomson commented:
This is great.

(Upon further reflection, I think that I may have led you on a wild goose chase w/r/t #2926...  I think that they're only tangentially related and I probably ended up making things less clear.  Apologies.)

Thanks for this, really appreciate it.  Hope WRI is well.
